launch/test/launch/utilities/test_type_utils.py

```diff
@@ -0,0 +1,578 @@
+from launch.utilities.type_utils import is_instance_of
+from launch.utilities.type_utils import is_instance_of_valid_type
+from launch.utilities.type_utils import is_normalized_substitution
+from launch.utilities.type_utils import is_substitution
+from launch.utilities.type_utils import is_typing_list
+from launch.utilities.type_utils import is_valid_scalar_type
+from launch.utilities.type_utils import normalize_typed_substitution
+from launch.utilities.type_utils import perform_typed_substitution
+
+import pytest
+
+
+def test_is_typing_list():
+    assert is_typing_list(List)
+    assert is_typing_list(List[int])
+    assert is_typing_list(List[float])
+    assert is_typing_list(List[Union[int, float]])
+    assert not is_typing_list(list)
+    assert not is_typing_list(int)
+    assert not is_typing_list(Union[int, float])
+    assert not is_typing_list(None)
+
+
+def test_is_valid_scalar_type():
+    assert is_valid_scalar_type(int)
+    assert is_valid_scalar_type(float)
+    assert is_valid_scalar_type(bool)
+    assert is_valid_scalar_type(str)
+    assert not is_valid_scalar_type(bytes)
+    assert not is_valid_scalar_type(None)
+    assert not is_valid_scalar_type(List[float])
+    assert not is_valid_scalar_type(list)
+    assert not is_typing_list(int)
+
+
+def test_extract_type():
+    assert extract_type(int) == (int, False)
+    assert extract_type(float) == (float, False)
+    assert extract_type(bool) == (bool, False)
+    assert extract_type(str) == (str, False)
+
+    assert extract_type(List[int]) == (int, True)
+    assert extract_type(List[float]) == (float, True)
+    assert extract_type(List[bool]) == (bool, True)
+    assert extract_type(List[str]) == (str, True)
+
+    with pytest.raises(ValueError):
+        extract_type(List)
+    with pytest.raises(ValueError):
+        extract_type(bytes)
+
+
+@pytest.mark.parametrize(
+    'is_instance_of_valid_type_impl',
+    (
+        is_instance_of_valid_type,
+        lambda x, y=False: is_instance_of(x, None, y),
+    ),
+    ids=[
+        'testing is_instance_of_valid_type implementation',
+        'testing is_instance_of implementation',
+    ]
+)
+def test_is_instance_of_valid_type(is_instance_of_valid_type_impl):
+    assert is_instance_of_valid_type_impl(1)
+    assert is_instance_of_valid_type_impl(1.)
+    assert is_instance_of_valid_type_impl('asd')
+    assert is_instance_of_valid_type_impl(True)
+
+    assert is_instance_of_valid_type_impl([1, 2])
+    assert is_instance_of_valid_type_impl([1., 2.])
+    assert is_instance_of_valid_type_impl(['asd', 'bsd'])
+    assert is_instance_of_valid_type_impl([True, False])
+
+    assert not is_instance_of_valid_type_impl([1, '2'])
+    assert not is_instance_of_valid_type_impl(object)
+    assert not is_instance_of_valid_type_impl(test_is_instance_of_valid_type)
+    assert not is_instance_of_valid_type_impl({'key': 'value'})
+
+    assert is_instance_of_valid_type_impl(1, True)
```

Comment:
@ivanpauno nit: here and below, using a `can_be_str` kwarg helps readability.

Reply:
done.
See https://github.com/ros2/launch/pull/438/commits/150deca7d9c9b37f030cce350450dadd0c623b0f and https://github.com/ros2/launch/pull/438/commits/99cc84c1e1d89ad706f2cbc2ddd3d1acc021ab98.